files/uncompress.py

```diff
@@ -0,0 +1,265 @@
+    """
+    Get the filetype from the unix command file, and then pick the correct compression method to use for it
+    """
+    cmd_path = module.get_bin_path('file')
+    cmd = "%s -b -i %s" % (cmd_path, src)
+    ftype = module.run_command(cmd, use_unsafe_shell=True)
+
+    return ftype
+
+def copyfile(src, dest, deep_check):
+    """
+    Copy file from tempsrc to final destination. Unless its already at dest, and the same as tempsrc.
+    """
+    changed = False
+    if os.path.isfile(dest):
+        # This takes a long time
+        if deep_check is True:
+            nodiff = filecmp.cmp(src, dest, shallow=True)
+        # This is much quicker
+        else:
+            destsize = os.path.getsize(dest)
+            srcsize  = os.path.getsize(src)
+            if destsize != srcsize:
+                nodiff = False
+            else:
+                nodiff = True
+        # If there is a difference, then we change the destination
+        if nodiff is False:
+            shutil.move(src, dest)
+            changed = True
+    # If the destination file does not exist, then place it.
+    else:
+        shutil.move(src, dest)
+        changed = True
+
+    return changed
+
+def main():
+    module = AnsibleModule(
+        # not checking because of daisy chain to file module
+        argument_spec = dict(
+            src               = dict(required=True),
+            dest              = dict(required=True),
+            copy              = dict(default=True, type='bool'),
+            original_basename = dict(required=False), # used to handle 'dest is a directory' via template, a slight hack
+            deep_check        = dict(default=False, type='bool'), # This check takes a long time if dest already exists.
+        ),
+        add_file_common_args=True,
+    )
+
+    src         = os.path.expanduser(module.params['src'])
+    dest        = os.path.expanduser(module.params['dest'])
+    copy        = module.params['copy']
+    deep_check  = module.params['deep_check']
+    file_args   = module.load_file_common_arguments(module.params)
+    tempdir     = os.path.dirname(__file__)
+    fdir, ffile = os.path.split(dest)
+
+    # did tar file arrive?
+    if not os.path.exists(src):
+        if copy:
+            module.fail_json(msg="Source '%s' failed to transfer" % src)
+        # If copy=false, and src= contains ://, try and download the file to a temp directory.
+        elif '://' in src:
+            package = os.path.join(tempdir, str(src.rsplit('/', 1)[1]))
+            try:
+                rsp, info = fetch_url(module, src)
+                f = open(package, 'w')
+                # Read 1kb at a time to save on ram
+                while True:
+                    data = rsp.read(BUFSIZE)
+
+                    if data == "":
+                        break # End of file, break while loop
+
+                    f.write(data)
+                f.close()
+                src = package
+            except Exception, e:
```

Comment:
There is no cleanup of the file if the download fail partially. IE, let's say that I start to download a 3G file and it break after 2G, I just have 2G of file in a directory which will not be cleaned by default (since that's ~/.ansible/tmp ).